hyp3lib/rtc2color.py

```diff
@@ -1,144 +1,147 @@
-      ['COMPRESS=LZW'])
-  outRaster.SetGeoTransform((originX, pixelWidth, 0, originY, 0, pixelHeight))
-  outRasterSRS = osr.SpatialReference()
-  outRasterSRS.ImportFromWkt(fullpol.GetProjectionRef())
-  outRaster.SetProjection(outRasterSRS.ExportToWkt())
-  fullpol = None
-  crosspol = None
-
-  print('Calculate red channel and save in GeoTIFF')
-  outBand = outRaster.GetRasterBand(1)
-  if real == True:
-    red = (2.0*rp*(1 - blue_mask) + zp*blue_mask)
-  else:
-    red = (2.0*rp*(1 - blue_mask) + zp*blue_mask)*255
-  red[red==0] = 1
-  red = red * mask
-
-  outBand.WriteArray(red)
-  red = None
-  print('Calculate green channel and save in GeoTIFF')
-  outBand = outRaster.GetRasterBand(2)
-  if real == True:
-    green = (3.0*np.sqrt(xp)*(1 - blue_mask) + 2.0*zp*blue_mask)
-  else:
-    green = (3.0*np.sqrt(xp)*(1 - blue_mask) + 2.0*zp*blue_mask)*255
-  green[green==0]=1
-  green = green * mask
-  outBand.WriteArray(green)
-  green = None
-  print('Calculate blue channel and save in GeoTIFF')
-  outBand = outRaster.GetRasterBand(3)
-  if real == True:
-    blue = (2.0*bp*(1 - blue_mask) + 5.0*zp*blue_mask)
-  else:
-    blue = (2.0*bp*(1 - blue_mask) + 5.0*zp*blue_mask)*255
-  blue[blue==0] = 1
-  blue = blue * mask
-  outBand.WriteArray(blue)
-  blue = None
-  xp = None
-  cp = None
-  zp = None
-  bp = None
-  blue_mask = None
-  outRaster = None
+def rtc2color(copol_tif: Union[str, Path], crosspol_tif: Union[str, Path], threshold: float, out_tif: Union[str, Path],
+              cleanup=False, teal=False, amp=False, real=False):
+    """RGB decomposition of a dual-pol RTC
+
+    Args:
+        copol_tif: The co-pol RTC GeoTIF
+        crosspol_tif: The cross-pol RTC GeoTIF
+        threshold: Decomposition threshold value in db
+        out_tif: The output color GeoTIFF file name
+        cleanup: Cleanup artifacts using a -48 db power threshold
+        teal: Combine green and blue channels because the volume to simple scattering ratio is high
+        amp: input TIFs are in amplitude and not power
+        real: Output real (floating point) values instead of RGB scaled (0--255) ints
+    """
+
+    # Suppress GDAL warnings but raise python exceptions
+    # https://gis.stackexchange.com/a/91393
+    gdal.UseExceptions()
+    gdal.PushErrorHandler('CPLQuietErrorHandler')
+
+    clean_threshold = pow(10.0, -48.0 / 10.0)  # db to power
+    power_threshold = pow(10.0, threshold / 10.0)  # db to power
+
+    # used scale the results to fit inside RGB 1-255 (ints), with 0 for no/bad data
+    scale_factor = 1.0 if real else 254.0
+    out_type = gdal.GDT_Float32 if real else gdal.GDT_Byte
+
+    copol = gdal.Open(copol_tif)
+    crosspol = gdal.Open(crosspol_tif)
+
+    geotransform = copol.GetGeoTransform()
+    proj_wkt = copol.GetProjectionRef()
+
+    cols = min(copol.RasterXSize, crosspol.RasterXSize)
+    rows = min(copol.RasterYSize, crosspol.RasterYSize)
+
+    logging.info(f'Reading co-pol image {copol_tif}')
```

Comment:
Reading co-pol and cross-pol are copy'paste of each other, consider a function, something like:
```
cp = read_image(copol, amp, cleanup, clean_threshold)
xp = read_image(crosspol, amp, cleanup, clean_threshold)
```

Reply:
I could do that -- I tend to follow the
* once: write
* twice: copy
* thrice: abstract 

"rule".

`rtc2colordiff.py` probably could utilize that function as well (and could use a refactor) so it's probably worth doing.